allennlp/models/coreference_resolution/coref.py

```diff
@@ -0,0 +1,411 @@
+    mention_feedforward : ``FeedForward``
+        This feedforward network is applied to the span representations which is then scored
+        by a linear layer.
+    antecedent_feedforward: ``FeedForward``
+        This feedforward network is applied to pairs of span representation, along with any
+        pairwise features, which is then scored by a linear layer.
+    feature_size: ``int``
+        The embedding size for all the embedded features, such as distances or span widths.
+    max_span_width: ``int``
+        The maximum width of candidate spans.
+    lexical_dropout: ``int``
+        The probability of dropping out dimensions of the embedded text.
+    initializer : ``InitializerApplicator``, optional (default=``InitializerApplicator()``)
+        Used to initialize the model parameters.
+    regularizer : ``RegularizerApplicator``, optional (default=``None``)
+        If provided, will be used to calculate the regularization penalty during training.
+    """
+    def __init__(self, vocab: Vocabulary,
+                 text_field_embedder: TextFieldEmbedder,
+                 context_layer: Seq2SeqEncoder,
+                 mention_feedforward: FeedForward,
+                 antecedent_feedforward: FeedForward,
+                 feature_size: int,
+                 max_span_width: int,
+                 spans_per_word: float,
+                 max_antecedents: int,
+                 lexical_dropout: float = 0.2,
+                 initializer: InitializerApplicator = InitializerApplicator(),
+                 regularizer: Optional[RegularizerApplicator] = None) -> None:
+        super(CoreferenceResolver, self).__init__(vocab, regularizer)
+        self._text_field_embedder = text_field_embedder
+        self._context_layer = context_layer
+        self._mention_feedforward = TimeDistributed(mention_feedforward)
+        self._antecedent_feedforward = TimeDistributed(antecedent_feedforward)
+        self._mention_scorer = TimeDistributed(torch.nn.Linear(mention_feedforward.get_output_dim(), 1))
+        self._antecedent_scorer = TimeDistributed(torch.nn.Linear(antecedent_feedforward.get_output_dim(), 1))
+        self._head_scorer = TimeDistributed(torch.nn.Linear(context_layer.get_output_dim(), 1))
+        self._distance_embedding = Embedding(10, feature_size) # 10 possible distance buckets.
+        self._span_width_embedding = Embedding(max_span_width, feature_size)
+        self._max_span_width = max_span_width
+        self._spans_per_word = spans_per_word
+        self._max_antecedents = max_antecedents
+        self._mention_recall = MentionRecall()
+        self._conll_coref_scores = ConllCorefScores()
+        if lexical_dropout > 0:
+            self._lexical_dropout = torch.nn.Dropout(p=lexical_dropout)
+        else:
+            self._lexical_dropout = lambda x: x
+        initializer(self)
+
+    def _compute_head_attention(self, head_scores, text_emb, span_ends, span_size):
```

Comment:
Nitty stuff now: Could you:
- Add types and docstrings to these private functions.
- use `embedding` rather than `emb` and `index` rather than `idx`